src/rgw/rgw_rados.cc

```diff
@@ -7679,6 +7682,14 @@ int RGWRados::follow_olh(const DoutPrefixProvider *dpp, RGWBucketInfo& bucket_in
     ldpp_dout(dpp, 20) << __func__ << "(): found pending entries, need to update_olh() on bucket=" << olh_obj->get_bucket() << dendl;
 
     int ret = update_olh(dpp, state, bucket_info, olh_obj);
+    if (ret == -ECANCELED) {
+      /* It's fine to ignore races because we only care that the BI OLH journal is eventually applied, not that it
+       * is applied by the current thread. Even if we were to retry in these cases, there could be another OLH
+       * change between the time that this function returns and the time that the caller uses the result. The caller
+       * must, therefore, always assume that the resolved OLH target could be stale.
+       */
+      ret = 0;
+    }
```

Comment:
in the context of these olh operations, `ECANCELED` means the "olh tag" changed - either on the 'olh object' in the data pool, or its associated 'olh entry' in the bucket index

my understanding of this "olh tag" is that it can only change if the old olh object and olh index entry are removed by the `if (need_to_remove) {` block at the end of [apply_olh_log()](https://github.com/ceph/ceph/blob/40dc7790/src/rgw/rgw_rados.cc#L7379-L7394). removal of these should only succeed if all prior versions of this object name had indeed been removed

so if `follow_olh()` continues on here after `ECANCELED` and uses this stale `RGWOLHInfo`, we'd expect that to always fail with `ENOENT` right? can we return `ENOENT` here directly to avoid potential rados reads?

Reply:
I agree with your understanding that the OLH tag will only change if the old OLH object/index were removed. It's somewhat difficult to reason about this precisely because errors are being "swallowed" all over the place in these code paths.

Using the call stack from my [comment below](https://github.com/ceph/ceph/pull/47187#issuecomment-1192836569), here's what I see in the case where `follow_olh()` continues after `ECANCELED` and uses the stale `RGWOLHInfo`:

- `follow_olh()` returns old target to `get_olh_target_state` invocation
- `get_obj_state` invocation within `get_olh_target_state()` returns 0 from [here](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_rados.cc#L5516)

And if we were to instead return ENOENT directly from `follow_olh()`, that would bubble all the way back up to the caller of `get_obj_state(follow_olh=true)`. This seems fine (and admittedly better) for the RGWDeleteMultiObj case since the ENOENT would cause some other code paths to be skipped and would ultimately result in the same DeleteObjects response, but I'm not sure if it's true for all callers of `get_obj_state(follow_olh=true)`

Reply:
Maybe we return ENOENT from `follow_olh()` but then still return 0 from `get_obj_state_impl()`?